logstash-core/src/main/java/org/logstash/config/ir/compiler/ComputeStepSyntaxElement.java

```diff
@@ -33,84 +32,98 @@
     private static final ISimpleCompiler COMPILER = new SimpleCompiler();
 
     /**
-     * Sequential counter to generate the class name
+     * Global cache of runtime compiled classes to prevent duplicate classes being compiled.
+     * across pipelines and workers.
      */
-    private static final AtomicLong classSeqCount = new AtomicLong();
+    private static final Map<ComputeStepSyntaxElement<?>, Class<? extends Dataset>> CLASS_CACHE
+        = new HashMap<>(5000);
 
     /**
      * Pattern to remove redundant {@code ;} from formatted code since {@link Formatter} does not
      * remove those.
      */
     private static final Pattern REDUNDANT_SEMICOLON = Pattern.compile("\n[ ]*;\n");
 
+    private static final String CLASS_NAME_PLACEHOLDER = "CLASS_NAME_PLACEHOLDER";
+
+    private static final Pattern CLASS_NAME_PLACEHOLDER_REGEX = Pattern.compile(CLASS_NAME_PLACEHOLDER);
+
     private final Iterable<MethodSyntaxElement> methods;
 
     private final ClassFields fields;
 
     private final Class<T> type;
 
-    private final long classSeq;
+    private final String normalizedSource;
 
     public static <T extends Dataset> ComputeStepSyntaxElement<T> create(
-        final Iterable<MethodSyntaxElement> methods, final ClassFields fields,
-        final Class<T> interfce) {
+        final Iterable<MethodSyntaxElement> methods,
+        final ClassFields fields,
+        final Class<T> interfce)
+    {
         return new ComputeStepSyntaxElement<>(methods, fields, interfce);
     }
 
-    private ComputeStepSyntaxElement(final Iterable<MethodSyntaxElement> methods,
-        final ClassFields fields, final Class<T> interfce) {
+    private ComputeStepSyntaxElement(
+        final Iterable<MethodSyntaxElement> methods,
+        final ClassFields fields,
+        final Class<T> interfce)
+    {
         this.methods = methods;
         this.fields = fields;
         type = interfce;
-        classSeq = classSeqCount.incrementAndGet();
+
+        // normalizes away the name of the class so that two classes of different name but otherwise
+        // equivalent syntax get correctly compared by {@link #equals(Object)}.
+        normalizedSource = generateCode(CLASS_NAME_PLACEHOLDER);
```

Comment:
:+1:

While I generally cringe at executing complex code like this in a constructor, the resulting value is intrinsically tied to the _identity_ of this object and deferring it to a separate method that we call out to in a synchronized in several places to make that work would only serve to obfuscate this fact.

Reply:
same same. I had it lazily initialized before but figured that the lazy initialization had no real value here and just added a bit of unnecessary code and complexity. 
As per you other comment this overall code-gen and compilation strategy should probably by  re evaluated which will result in probably decoupling some of that code here.